File: CPP/059_1.cpp
#include <iostream>
#include <cstdint>
#include <vector>

#include <boost/range/irange.hpp>
#include <boost/range/algorithm/fill.hpp>

using namespace std;

int main() {
	uint N, M, Q;
	cin >> N >> M >> Q;

	vector<vector<uint>> adjacency(N);

	for (uint i : boost::irange(M)) {
		uint x, y;
		cin >> x >> y;
		adjacency[x - 1].push_back(y - 1);
	}

	vector<uint64_t> bits(N);
	vector<uint> goals;
	goals.reserve(64);

	while (Q > 0) {
		boost::fill(bits, 0UL);
		goals.clear();

		uint64_t bit = 1UL;
		while (bit > 0 && Q > 0) {
			uint a, b;
			cin >> a >> b;
			bits[a - 1] |= bit;
			goals.push_back(b - 1);

			bit <<= 1;
			Q--;
		}

		for (uint i : boost::irange(N)) {
			bit = bits[i];
			for (uint neighbor : adjacency[i]) {
				bits[neighbor] |= bit;
			}
		}

		bit = 1UL;
		for (uint goal : goals) {
			cout << (bits[goal] & bit ? "Yes" : "No") << endl;
			bit <<= 1;
		}
	}

	return 0;
}

File: CSharp/059_1.cs
#nullable enable

using System;
using System.Collections.Generic;

using static System.Console;
//using static System.Math
using static System.Linq.Enumerable;

public class Program
{
	public static void Main()
	{
		string[] s = ReadLine()!.Split();
		uint N = uint.Parse(s[0]);
		uint M = uint.Parse(s[1]);
		uint Q = uint.Parse(s[2]);

		List<uint>[] adjacency_lists = Range(0, (int)N)
			.Select(i => new List<uint>()).ToArray();

		foreach (int _ in Range(0, (int)M)) {
			s = ReadLine()!.Split();
			uint x = uint.Parse(s[0]) - 1;
			uint y = uint.Parse(s[1]) - 1;
			adjacency_lists[x].Add(y);
		}

		uint[][] adjacency = adjacency_lists.Select(adj => adj.ToArray()).ToArray();

		var bits = new ulong[N];
		var goals = new List<uint>(64);

		while (Q > 0) {
			Array.Fill(bits, 0UL);
			goals.Clear();

			ulong bit = 1UL;
			while (bit > 0 && Q > 0) {
				s = ReadLine()!.Split();
				uint a = uint.Parse(s[0]) - 1;
				uint b = uint.Parse(s[1]) - 1;
				bits[a] |= bit;
				goals.Add(b);

				bit <<= 1;
				Q--;
			}

			for (uint i = 0; i < N; i++) {
				bit = bits[i];
				foreach (uint neighbor in adjacency[i]) {
					bits[neighbor] |= bit;
				}
			}

			bit = 1;
			foreach (uint goal in goals) {
				WriteLine((bits[goal] & bit) != 0 ? "Yes" : "No");
				bit <<= 1;
			}
		}
	}
}
